vendor/tu_collector/tu_collector/tu_collector.py

```diff
@@ -0,0 +1,312 @@
+        for f in files:
+            archive_path = os.path.join('sources-root', f.lstrip(os.sep))
+
+            try:
+                archive.getinfo(archive_path)
+            except KeyError:
+                archive.write(f, archive_path, zipfile.ZIP_DEFLATED)
+            else:
+                logging.debug("'%s' is already in the ZIP file, won't add it "
+                              "again!", f)
+
+
+def zip_tu_files(zip_file, compilation_database, write_mode='w'):
+    """
+    Collects all files to a zip file which are required for the compilation of
+    the translation units described by the given compilation database.
+    If there are some files which couldn't be collected then a "no-sources"
+    file will contain the error message of its reason.
+    The function returns the set of files.
+
+    zip_file -- A file name or a file object.
+    compilation_database -- Either a path of the compilation database JSON file
+                            or a list of the parsed JSON.
+    write_mode -- The file opening mode of the zip_file. In case of 'a' the new
+                  files are appended to the existing zip file, in case of 'w'
+                  the files are added to a clean zip file.
+    """
+    if isinstance(compilation_database, basestring):
+        with open(compilation_database) as f:
+            compilation_database = json.load(f)
+
+    no_sources = 'no-sources'
+    tu_files = set()
+    error_messages = ''
+
+    for buildaction in compilation_database:
+        files, err = get_dependent_headers(
+            buildaction['command'],
+            buildaction['directory'])
+
+        tu_files |= files
+
+        if err:
+            error_messages += buildaction['file'] + '\n' \
+                + '-' * len(buildaction['file']) + '\n' + err + '\n'
+
+    if write_mode == 'a' and os.path.isfile(zip_file):
+        with zipfile.ZipFile(zip_file) as archive:
+            try:
+                archive.getinfo(no_sources)
+            except KeyError:
+                pass
+            else:
+                no_sources = 'no-sources_' + __random_string(5)
+
+    if error_messages:
+        with zipfile.ZipFile(zip_file, write_mode) as archive:
+            archive.writestr(no_sources, error_messages)
+
+    add_sources_to_zip(zip_file, tu_files)
+
+
+def main():
+    # --- Handling of command line arguments --- #
+
+    parser = argparse.ArgumentParser(
+        description='Collect files of translation units.')
+
+    parser.add_argument('-z', '--zip', dest='zip', type=str, required=True,
+                        help="Output ZIP file.")
+    parser.add_argument('-c', '--compilation-db', dest='compilation_db',
+                        type=str, required=True,
+                        help="Compilation database JSON file.")
+    parser.add_argument('-f', '--filter', dest='filter',
+                        type=str, required=False,
+                        help="This flag has to be given a file.")
```

Comment:
Please add some examples here what are the valid filter values. (main.cpp, ma*.cpp, */mylib/lib.c)